dhcpv4/nclient4/lease.go

```diff
@@ -0,0 +1,227 @@
+		return
+	}
+	//save lease
+	c.lease = &DHCPv4ClientLease{}
+	c.lease.AssignedIP = ack.YourIPAddr
+	c.lease.AssignedIPMask = ack.SubnetMask()
+	c.lease.CreationTime = time.Now()
+	c.lease.LeaseDuration = ack.IPAddressLeaseTime(0)
+	if c.lease.LeaseDuration == 0 {
+		c.lease.LeaseDuration = defaultLeaseTime
+		c.logger.Printf("warning: server doesn't include Lease Time option or it is zero seconds, setting lease time to default %v", c.lease.LeaseDuration)
+
+	}
+	c.lease.RenewInterval = ack.IPAddressRenewalTime(0)
+	if c.lease.RenewInterval == 0 {
+		//setting default to half of lease time based on RFC2131,section 4.4.5
+		c.lease.RenewInterval = time.Duration(float64(c.lease.LeaseDuration) / 2)
+		c.logger.Printf("warning: server doesn't include Renew Time option or it is zero seconds, setting lease time to default %v", c.lease.RenewInterval)
+
+	}
+	c.lease.RebindInterval = ack.IPAddressRebindingTime(0)
+	if c.lease.RebindInterval == 0 {
+		//setting default to 0.875 of lease time based on RFC2131,section 4.4.5
+		c.lease.RebindInterval = time.Duration(float64(c.lease.LeaseDuration) * 0.875)
+		c.logger.Printf("warning: server doesn't include Renew Time option or it is zero seconds, setting lease time to default %v", c.lease.RebindInterval)
+
+	}
+	c.lease.IDOptions = dhcpv4.Options{}
+	for _, optioncode := range c.clientIDOptions {
+		v := request.Options.Get(optioncode)
+		c.lease.IDOptions.Update(dhcpv4.OptGeneric(optioncode, v))
+	}
+	c.lease.AckOptions = ack.Options
+	//update server address
+	c.serverAddr = &(net.UDPAddr{IP: ack.ServerIdentifier(), Port: 67})
+	err = c.ApplyLease(true)
+	return
+}
+
+//Release send DHCPv4 release messsage to server.
+//release is sent as unicast per RFC2131, section 4.4.4.
+//The lease need to be applied with c.ApplyLease(true) first before calling Release.
+func (c *Client) Release() error {
+	if c.lease == nil {
+		return fmt.Errorf("There is no lease to release")
+	}
+	req, err := dhcpv4.New()
+	if err != nil {
+		return err
+	}
+	//This is to make sure use same client identification options used during
+	//DORA, so that DHCP server could identify the required lease
+	req.Options = c.lease.IDOptions
+
+	req.UpdateOption(dhcpv4.OptMessageType(dhcpv4.MessageTypeRelease))
+	req.ClientHWAddr = c.ifaceHWAddr
+	req.ClientIPAddr = c.lease.AssignedIP
+	req.UpdateOption(dhcpv4.OptServerIdentifier(c.serverAddr.IP))
+	req.SetUnicast()
+	luaddr, err := net.ResolveUDPAddr("udp4", fmt.Sprintf("%v:%v", c.lease.AssignedIP, 68))
+	if err != nil {
+		return err
+	}
+
+	uniconn, err := net.DialUDP("udp4", luaddr, c.serverAddr)
+	if err != nil {
+		return err
+	}
+	_, err = uniconn.Write(req.ToBytes())
+	if err != nil {
+		return err
+	}
+	c.logger.PrintMessage("sent message:", req)
+	return c.ApplyLease(false)
+}
+
+//NewWithLease return a Client with populated lease.
+//this function could be used to release a saved lease.
+func NewWithLease(clease DHCPv4ClientLease, opts ...ClientOpt) (*Client, error) {
```

Comment:
This actually sort of illustrates why I'd rather the lease not be stored in the *Client object -- now you have to construct a Client with a lease to be able to release it.

Reply:
no, NewWithLease is only for use case like client did DORA, save the lease to a file, and quit running, then after some time, start again, read lease from the file, create a client with saved lease, and could do renew/release with the loaded lease;
if you don't need such use case, you could simply call Client.Release() after DORA, no need to create a new client.

Reply:
What if I have a client with more than one lease to more than one DHCP server?

Reply:
current design of this PR assume only one lease per client, for multiple leases, then multiple clients need to be created; which I think is acceptable design, because there could be significant differences between leases (e.g different binding interface), so there is not much benefit to share a single client instance

Reply:
I was thinking more of multiple leases per interface.

As I said elsewhere, I'm happy to have a more stateful DHCP client in a different package in this library, but I'd rather keep nclient4 a little less stateful. We use this DHCP client for netbooting mostly, and in that case we don't need and don't want this state to be around, so it'd be great to have a way to separate the use cases.